### PR TITLE
convert to mpp structure

### PR DESCRIPTION
**Pull request by Munzey**
Proposed solution for #12
Things that had to change:
- moved files to adhere to app structure by placing in commonMain and commonTest
- `Binding.kt` overriding stack trace can’t be used in kotlin native as exception doesn’t have this function. **_EDIT:_** now using expect/actual keywords with jvm implementation that still overrides stack trace
- ~`BindingTest.kt `was importing junit test instead of kotlin test~ **_EDIT:_** merged in separate pr #19
- example application now needed to specify that it was compiling against Jvm 1.8
- added a jvmBenchmark sourceset for benchmarks that required java plugin to be enabled for the mpp jvm target

### CODE REVIEW

File: build.gradle.kts [hunk 1]
```diff
@@ -10,86 +9,115 @@ description = "A Result monad for modelling success or failure operations."
...
+}
+
 allprojects {
     repositories {
         mavenCentral()
         jcenter()
         maven("https://dl.bintray.com/kotlin/kotlinx")
     }
-
-    tasks.withType<KotlinCompile> {
```
Comment (Munzey) [inline build.gradle.kts:28]: this is now handled via language settings in mpp kotlin extension below

File: build.gradle.kts [hunk 2]
```diff
@@ -10,86 +9,115 @@ description = "A Result monad for modelling success or failure operations."
...
+        val jvmMain by getting {
+            dependencies {
+                implementation(kotlin("stdlib-jdk8"))
+            }
+        }
+
+        val jvmTest by getting {
+            dependencies {
+                implementation(kotlin("test-junit"))
+                implementation(kotlin("test"))
```
Comment (Munzey) [inline build.gradle.kts:91]: should this be moved to commonTest?

File: build.gradle.kts [hunk 3]
```diff
@@ -10,86 +9,115 @@ description = "A Result monad for modelling success or failure operations."
...
     repositories {
         mavenCentral()
         jcenter()
         maven("https://dl.bintray.com/kotlin/kotlinx")
     }
-
-    tasks.withType<KotlinCompile> {
-        kotlinOptions {
-            jvmTarget = "1.8"
-            freeCompilerArgs = listOf("-Xuse-experimental=kotlin.contracts.ExperimentalContracts")
```
Comment (Munzey) [inline build.gradle.kts:31]: this is now handled via language settings in mpp kotlin extension below

File: src/commonMain/kotlin/com/github/michaelbull/result/Binding.kt [hunk 4]
```diff
@@ -38,11 +38,7 @@ inline fun <V, E> binding(crossinline block: ResultBinding<E>.() -> V): Result<V
     }
 }
 
-internal object BindException : Exception() {
-    override fun fillInStackTrace(): Throwable {
-        return this
-    }
-}
```
Comment (michaelbull) [inline src/commonMain/kotlin/com/github/michaelbull/result/Binding.kt:45]: Given this change, surely we need to change this to an `expect` and `actual` contract, i.e. `expect class BindException` to exist and on the JVM one explicitly override the `fillInStackTrace()` in an `actual class BindException`? Otherwise we lose the behaviour of stacktrace avoidance

File: build.gradle.kts [hunk 5]
```diff
@@ -10,86 +9,115 @@ description = "A Result monad for modelling success or failure operations."
...
 allprojects {
     repositories {
         mavenCentral()
         jcenter()
         maven("https://dl.bintray.com/kotlin/kotlinx")
     }
-
-    tasks.withType<KotlinCompile> {
-        kotlinOptions {
-            jvmTarget = "1.8"
```
Comment (Munzey) [inline build.gradle.kts:30]: this is now specified via `jvm().compilations.all` below

File: src/commonTest/kotlin/com/github/michaelbull/result/BindingTest.kt [hunk 6]
```diff
@@ -1,6 +1,6 @@
 package com.github.michaelbull.result
 
-import org.junit.Test
```
Comment (michaelbull) [inline src/commonTest/kotlin/com/github/michaelbull/result/BindingTest.kt:3]: Can we split this change up into a separate PR? This can be done in isolation to the general MPP migration and is a good change to have whether the mpp gets merged or not.

File: src/commonMain/kotlin/com/github/michaelbull/result/Binding.kt (hunk 4 above)
Comment (Munzey) [reply src/commonMain/kotlin/com/github/michaelbull/result/Binding.kt:45]: ah good thinking!

File: build.gradle.kts [hunk 7]
```diff
@@ -10,86 +9,115 @@ description = "A Result monad for modelling success or failure operations."
...
-    testImplementation(kotlin("test-common"))
-    testImplementation(kotlin("test-annotations-common"))
-    testImplementation(kotlin("test-junit"))
-    testImplementation(kotlin("test"))
+kotlin {
+    jvm {
+        withJava()
+    }
+    sourceSets {
+        val commonMain by getting {
```
Comment (michaelbull) [inline build.gradle.kts:69]: Can we make this use the `named` methods, just like I do here: https://github.com/michaelbull/kotlin-inline-logger/blob/master/build.gradle.kts#L46

Otherwise IntelliJ moans that these `val`s (commonMain, commonTest, etc) are unused variables.

Comment (Munzey) [reply build.gradle.kts:69]: ill take a look, it made specifying that `jvmBenchmark dependsOn(jvmMain)` nice and readable

File: build.gradle.kts (hunk 2 above)
Comment (michaelbull) [reply build.gradle.kts:91]: I don't think so, see: https://github.com/michaelbull/kotlin-inline-logger/blob/master/build.gradle.kts#L73

File: src/commonTest/kotlin/com/github/michaelbull/result/BindingTest.kt (hunk 6 above)
Comment (Munzey) [reply src/commonTest/kotlin/com/github/michaelbull/result/BindingTest.kt:3]: sure thats a fair point!

File: build.gradle.kts [hunk 8]
```diff
@@ -10,86 +9,115 @@ description = "A Result monad for modelling success or failure operations."
...
-dependencies {
-    implementation(kotlin("stdlib"))
-    testImplementation("junit:junit:4.13")
-    testImplementation(kotlin("test-common"))
-    testImplementation(kotlin("test-annotations-common"))
-    testImplementation(kotlin("test-junit"))
-    testImplementation(kotlin("test"))
+kotlin {
+    jvm {
+        withJava()
```
Comment (michaelbull) [inline build.gradle.kts:66]: What does this withJava thing do? I don't remember needing it in my other mpp library

File: build.gradle.kts [hunk 9]
```diff
@@ -10,86 +9,115 @@ description = "A Result monad for modelling success or failure operations."
...
+            artifact(javadocJar.get())
+        }
+    }
 
-val sourcesJar by tasks.registering(Jar::class) {
-    group = LifecycleBasePlugin.BUILD_GROUP
-    description = "Assembles a jar archive containing the main classes with sources."
-    archiveClassifier.set("sources")
-    from(sourceSets["main"].allSource)
+    sourceSets.all {
```
Comment (michaelbull) [inline build.gradle.kts:116]: Surely this can be put inside the other `sourceSets` block?

```kotlin
sourceSets {
    all {
         ...
    }

    named("commonMain") {
        ...
    }

...
...
```

File: build.gradle.kts (hunk 7 above)
Comment (michaelbull) [reply build.gradle.kts:69]: Oh if you are actually using the `jvmMain` as a variable then ignore me!

File: build.gradle.kts (hunk 8 above)
Comment (Munzey) [reply build.gradle.kts:66]: this was the magic line that got benchmarking working, see my comment here: https://github.com/michaelbull/kotlin-result/pull/18/files#diff-392475fdf2bc320d17762ed97109a121R54-R55

Comment (michaelbull) [reply build.gradle.kts:66]: Haha, I can see why I couldn't get benchmarking working then... nice find!

Comment (Munzey) [reply build.gradle.kts:66]: without it, we don't get a target that the benchmark plugin will recognize

File: build.gradle.kts (hunk 9 above)
Comment (Munzey) [reply build.gradle.kts:116]: makes sense! 👍

File: src/commonMain/kotlin/com/github/michaelbull/result/Binding.kt (hunk 4 above)
Comment (Munzey) [reply src/commonMain/kotlin/com/github/michaelbull/result/Binding.kt:45]: done

File: build.gradle.kts (hunk 9 above)
Comment (Munzey) [reply build.gradle.kts]: done

File: build.gradle.kts [hunk 10]
```diff
@@ -10,86 +9,115 @@ description = "A Result monad for modelling success or failure operations."
...
+        kotlinOptions {
+            jvmTarget = "1.8"
+        }
+    }
+
+    jvm {
+        mavenPublication {
+            artifact(javadocJar.get())
+        }
+    }
```
Comment (michaelbull) [inline build.gradle.kts:120]: Can these two jvm blocks be merged?

Something like

```kotlin
    jvm {
        mavenPublication {
            artifact(javadocJar.get())
        }

        compilations.all {
            kotlinOptions.jvmTarget = "1.8"
        }
    }
```

Comment (Munzey) [reply build.gradle.kts:120]: yes! ill move them into the same block as `withJava()`

Comment (Munzey) [reply build.gradle.kts:118]: done

File: src/commonTest/kotlin/com/github/michaelbull/result/BindingTest.kt (hunk 6 above)
Comment (Munzey) [reply src/commonTest/kotlin/com/github/michaelbull/result/BindingTest.kt:3]: opened pr: https://github.com/michaelbull/kotlin-result/pull/19

**Comment by michaelbull** [comment]
Hi @Munzey, don't want to leave you in silence but the PR looks good right now. I'm busy right now so will hopefully get some time this weekend to merge it and release. I tried last weekend to release the binding behaviour but I had connection issues with uploading artifacts to maven central. Apologies for trailing on this one.

**Comment by Munzey** [comment]
@michaelbull No worries! 
Appreciate the update :relaxed:

**Comment by michaelbull** [comment]
Hi @Munzey. Been thinking about this change a bit today. Given we introduced binding before MPP but now the binding is platform dependent (expects a platform-specific impl of that Exception object we spoke about), I guess that means that we would need to add an equivalent implementation for every platform we support - whereas before the binding stuff we could have technically supported every platform as none of the code was platform specific in any way. I may be misunderstanding this though?

I was also therefore thinking of splitting out the binding stuff into a separate subproject and therefore separate artifact, such that the main library is platform independent and the binding stuff is platform dependent. I'm not sure how I feel about this and it could get ugly pretty quickly, maybe I'm misunderstanding the situation though. Let me know your thoughts.

**Comment by Munzey** [comment]
>whereas before the binding stuff we could have technically supported every platform as none of the code was platform specific in any way. I may be misunderstanding this though?

@michaelbull I think you are right yes.

>I was also therefore thinking of splitting out the binding stuff into a separate subproject and therefore separate artifact, such that the main library is platform independent and the binding stuff is platform dependent. I'm not sure how I feel about this and it could get ugly pretty quickly,

This does sound like it would become a pain to support both. Some questions around it too, would it be two dependencies to add to a project like `mpp.result-core` and `jvm.result-binding`? or is there a `mpp.result-core` and a `jvm.result-extended` that comprises everything? Looking at a project like arrow, I think it can become confusing when there is multiple artifacts required to pull all with names that seem to suggest they do something similar 😅 .

Its a shame we can't have an expected `BindException` with a default implementation in the common code. There is actually an open issue to add this to kotlin (which I will now be up voting haha): https://youtrack.jetbrains.com/issue/KT-20427?_ga=2.19951672.1617774961.1592496293-2122254457.1527159951

The alternative here is to just not override the stacktrace for the jvm, I decided to look for some articles to mention how much faster is it to not fill in the stack trace. I did find one: https://www.atlassian.com/blog/archives/if_you_use_exceptions_for_path_control_dont_fill_in_the_stac
that suggests it is quite beneficial when used as part of flow control as it is here. Not sure if the current benchmarking we have written could be used to give much of an indicator.

Going back to the original open issue, I actually wonder is their even added value right now for making the project mpp to support android. The only thing preventing me pulling this lib into an android project was to specify that I wanted kotlin to compile against 1.8. Thats all it took. So It doesn't really need to be an mpp project to support android as far as I understand (I would assume for a kotlin native project that builds for both iOS and android that you could specify deps that are only for the android side of code).

That then leaves mpp projects where kotlin code is to be cross compiled to run for that platform , which right now I can imagine would be JS or Swift. Which comes back to your point
>I guess that means that we would need to add an equivalent implementation for every platform we support 

I think its really up to you whether you think its worth it 😄 Personally I'm very much in favor of keeping binding as part of the main lib, and simply adding the cross platform support for languages as the issues are opened. To hedge our bets I would say if we added a swift `actual` and a js `actual` and added those to the gradle mpp builds, there probably wouldnt be many requests coming in.

TL;DR im not sure lol

**Comment by avently** [comment]
Everyday I open this page to look at the status of merging process. That's because I really need functionality that this library gives but I can't use it without multiplatform support. So i'm just wondering, will this pr be merged in near future or it can be unmerged for months?

**Comment by Munzey** [comment]
@avently can I ask what platforms you plan to support with your project?

**Comment by avently** [comment]
@Munzey Android and Java right now, Linux/Windows/Mac/iOS later.

**Comment by michaelbull** [comment]
Merged in bca344daafadbc7af42089b88c712ac1061298dd

**Comment by avently** [comment]
@Munzey and @michaelbull thank you, guys!
